cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -85,26 +85,44 @@ void OptimizationProblem::AddTrajectoryNode(
   CHECK_GE(trajectory_id, 0);
   node_data_.resize(
       std::max(node_data_.size(), static_cast<size_t>(trajectory_id) + 1));
-  node_data_[trajectory_id].push_back(
-      NodeData{time, initial_point_cloud_pose, point_cloud_pose});
   trajectory_data_.resize(std::max(trajectory_data_.size(), node_data_.size()));
+
+  auto& trajectory_data = trajectory_data_.at(trajectory_id);
+  node_data_[trajectory_id].emplace(
+      trajectory_data.next_node_index,
+      NodeData{time, initial_point_cloud_pose, point_cloud_pose});
+  ++trajectory_data.next_node_index;
 }
 
 void OptimizationProblem::TrimTrajectoryNode(const mapping::NodeId& node_id) {
-  auto& trajectory_data = trajectory_data_.at(node_id.trajectory_id);
-  // We only allow trimming from the start.
-  CHECK_EQ(trajectory_data.num_trimmed_nodes, node_id.node_index);
   auto& node_data = node_data_.at(node_id.trajectory_id);
-  CHECK(!node_data.empty());
   if (node_id.trajectory_id < static_cast<int>(imu_data_.size())) {
-    const common::Time node_time = node_data.front().time;
-    auto& imu_data = imu_data_.at(node_id.trajectory_id);
-    while (imu_data.size() > 1 && imu_data[1].time <= node_time) {
-      imu_data.pop_front();
+    auto node_it = node_data.begin();
+    for (; std::next(node_it, 1) != node_data.end(); ++node_it) {
+      if (std::next(node_it, 1)->first == node_id.node_index) {
```

Comment:
This seems a complicated way to implement `find(node_id.node_index)`. Most importantly for trimming the first node of trajectory (which is what we do right now), it seems to never trim any IMU data. Rewrite? Maybe along these lines:

- Remove node from `node_data`.
- If the trajectory is not yet empty, find the time of the now first node and call it `node_time`.
- following logic as before

Does that make sense?

Reply:
I don't see how it deletes imu_date in the middle. But, it should be fine since there is only `PureLocalizationTrimmer`. Ok, I will follow the suggestion.

Reply:
It does not and it might make sense to add this functionality in the future. I just wanted to make sure there is no regression.